src/BinaryParsers/ElfBinary/Dwarf/DwarfLineNumberProgram.cs

```diff
@@ -173,62 +178,212 @@
         /// <param name="debugLine">The debug line data stream.</param>
         /// <param name="addressNormalizer">Normalize address delegate (<see cref="NormalizeAddressDelegate"/>)</param>
         /// <returns>List of file information.</returns>
-        private static List<DwarfFileInformation> ReadData(DwarfMemoryReader debugLine, NormalizeAddressDelegate addressNormalizer)
+        private static List<DwarfFileInformation> ReadData(int dwarfVersion,
+                                                           DwarfMemoryReader debugLine,
+                                                           DwarfMemoryReader debugStrings,
+                                                           DwarfMemoryReader debugLineStrings,
+                                                           NormalizeAddressDelegate addressNormalizer)
         {
             // Read header
-            ulong length = debugLine.ReadLength(out bool is64bit);
-            int endPosition = debugLine.Position + (int)length;
+            ulong unitLength = debugLine.ReadLength(out bool is64bit);
+            if (unitLength <= 1)
+            {
+                // A null condition is a clue to the upstream parser that it shouldn't
+                // continue, as we're in an invalid condition. We hit this particular 
+                // code path parsing binaries in the wild with language MipsAssembler.
+                return null;
+            }
+
+            int endPosition = debugLine.Position + (int)unitLength;
 
             if (endPosition > debugLine.Data.Length - 1)
             {
-                endPosition = debugLine.Data.Length - 1;
+                return null;
             }
 
-            debugLine.ReadUshort(); // version
-            debugLine.ReadOffset(is64bit); // headerLength
+            ushort version = debugLine.ReadUshort();
+
+            // DWARF versions prior to v2 did not have a .debug_line section, so
+            // this is a condition that indicates a parsing issue or invalid
+            // or corrupt binary.
+            if (version < 2)
+            {
+                return null;
+            }
+
+            if (dwarfVersion == 5)
+            {
+                byte addressSize = debugLine.ReadByte();
+                byte segmentSelectorSize = debugLine.ReadByte();
+            }
+
+            int headerLength = debugLine.ReadOffset(is64bit);
             byte minimumInstructionLength = debugLine.ReadByte();
-            bool defaultIsStatement = debugLine.ReadByte() != 0;
+            byte maximumOperationsPerInstruction = debugLine.ReadByte();
+
+            bool defaultIsStatement = debugLine.ReadByte() == 1;
```

Comment:
these 2 line change we now `debugLine.ReadByte()` one more time than before, was it a bug before?

Reply:
Nice catch! This data was introduced in V4, updated code.